File: generate_contributions.js
#!/usr/bin/env node
/**
 * Generates realistic-looking GitHub contribution history via backdated git commits.
 * Creates a pattern that looks like a real final-year CS student.
 */

const { execSync } = require('child_process');
const fs = require('fs');
const path = require('path');

const REPO_DIR = __dirname;
const START_DATE = new Date(2025, 6, 1); // July 1, 2025 (months are 0-indexed in JS)
const END_DATE = new Date(2026, 6, 20);  // July 20, 2026

// Seasonal activity weights
const MONTH_WEIGHTS = {
    0: 0.30, 1: 0.50, 2: 0.65, 3: 0.70, 4: 0.35, 5: 0.25,
    6: 0.55, 7: 0.60, 8: 0.70, 9: 0.75, 10: 0.65, 11: 0.30
};

// Day-of-week weights (0=Sunday, 1=Monday... JS getDay)
const DAY_WEIGHTS = {
    1: 0.80, 2: 0.85, 3: 0.80, 4: 0.75, 5: 0.65, 6: 0.35, 0: 0.20
};

const BURST_PERIODS = [
    [new Date(2025, 9, 10), new Date(2025, 9, 18)],
    [new Date(2025, 10, 20), new Date(2025, 10, 30)],
    [new Date(2026, 1, 5), new Date(2026, 1, 15)],
    [new Date(2026, 2, 15), new Date(2026, 2, 25)],
    [new Date(2026, 3, 10), new Date(2026, 3, 20)],
    [new Date(2026, 6, 1), new Date(2026, 6, 15)]
];

const DEAD_PERIODS = [
    [new Date(2025, 11, 1), new Date(2025, 11, 20)],
    [new Date(2025, 11, 24), new Date(2026, 0, 3)],
    [new Date(2026, 0, 5), new Date(2026, 0, 20)],
    [new Date(2026, 4, 1), new Date(2026, 4, 20)],
    [new Date(2026, 5, 5), new Date(2026, 5, 18)]
];

const COMMIT_MESSAGES = [
    "refactor: clean up utility functions", "feat: add input validation",
    "fix: resolve edge case in parser", "docs: update API documentation",
    "style: improve code formatting", "test: add unit tests for auth module",
    "chore: update dependencies", "feat: implement caching layer",
    "fix: handle null pointer exception", "refactor: extract helper methods",
    "feat: add error handling middleware", "docs: add inline comments",
    "fix: correct off-by-one error", "feat: implement pagination",
    "chore: clean up unused imports", "test: add integration tests",
    "feat: add logging infrastructure", "fix: resolve race condition",
    "refactor: optimize database queries", "feat: implement rate limiting",
    "docs: update README with examples", "fix: handle timeout errors gracefully",
    "feat: add configuration validation", "style: standardize naming conventions",
    "test: improve test coverage", "feat: add health check endpoint"
];

const FILE_CONTENTS = [
    "// Updated: {date}\n// Module improvements\nconst VERSION = '{version}';\n",
    "# Config updated {date}\nDEBUG={debug}\nLOG_LEVEL=info\n",
    "/* Style update {date} */\n.container {{ margin: {margin}px; }}\n",
    "# Notes - {date}\n- Refactored module\n- Fixed {bugs} bugs\n",
    "// Test suite - {date}\ndescribe('module', () => {{ /* tests */ }});\n"
];

function isInPeriod(date, periods) {
    return periods.some(([start, end]) => date >= start && date <= end);
}

function randomChoice(arr) {
    return arr[Math.floor(Math.random() * arr.length)];
}

function randomInt(min, max) {
    return Math.floor(Math.random() * (max - min + 1)) + min;
}

function getCommitCount(date) {
    if (isInPeriod(date, DEAD_PERIODS)) return 0;

    let monthWeight = MONTH_WEIGHTS[date.getMonth()];
    let dayWeight = DAY_WEIGHTS[date.getDay()];

    if (isInPeriod(date, BURST_PERIODS)) {
        monthWeight = Math.min(monthWeight * 1.8, 1.0);
        dayWeight = Math.min(dayWeight * 1.5, 1.0);
    }

    let probability = monthWeight * dayWeight;
    probability *= (Math.random() * (1.4 - 0.6) + 0.6);

    if (Math.random() > probability) return 0;

    if (isInPeriod(date, BURST_PERIODS)) {
        // Higher intensity 1-7
        const weights = [15, 25, 25, 15, 10, 5, 5];
        return weightedRandom([1, 2, 3, 4, 5, 6, 7], weights);
    } else {
        // Normal 1-4
        const weights = [45, 30, 15, 10];
        return weightedRandom([1, 2, 3, 4], weights);
    }
}

function weightedRandom(items, weights) {
    let i;
    let sum = weights.reduce((a, b) => a + b, 0);
    let rand = Math.random() * sum;
    for (i = 0; i < items.length; i++) {
        rand -= weights[i];
        if (rand < 0) return items[i];
    }
    return items[0];
}

function createCommit(date, index) {
    const hours = [
        ...Array(6).fill(5).map((w, i) => [i, w]),
        ...Array(3).fill(5).map((w, i) => [i+6, w]),
        ...Array(3).fill(15).map((w, i) => [i+9, w]),
        ...Array(2).fill(10).map((w, i) => [i+12, w]),
        ...Array(4).fill(25).map((w, i) => [i+14, w]),
        ...Array(4).fill(30).map((w, i) => [i+18, w]),
        ...Array(2).fill(10).map((w, i) => [i+22, w])
    ];
    
    const hList = hours.map(h => h[0]);
    const wList = hours.map(h => h[1]);
    
    let hour = weightedRandom(hList, wList);
    let minute = randomInt(0, 59);
    let second = randomInt(0, 59);

    let d = new Date(date);
    d.setHours(hour, minute, second);
    
    let pad = (n) => String(n).padStart(2, '0');
    let dateStr = `${d.getFullYear()}-${pad(d.getMonth()+1)}-${pad(d.getDate())}T${pad(d.getHours())}:${pad(d.getMinutes())}:${pad(d.getSeconds())}`;

    let message = randomChoice(COMMIT_MESSAGES);
    
    let contentTpl = randomChoice(FILE_CONTENTS);
    let dateFmt = `${d.getFullYear()}-${pad(d.getMonth()+1)}-${pad(d.getDate())}`;
    
    let content = contentTpl
        .replace('{date}', dateFmt)
        .replace('{version}', `${randomInt(1,9)}.${randomInt(0,9)}.${randomInt(0,99)}`)
        .replace('{debug}', randomChoice(['true', 'false']))
        .replace('{margin}', randomInt(4, 32))
        .replace('{bugs}', randomInt(1, 12))
        .replace('{{', '{').replace('}}', '}');

    const filenames = [
        "src/utils.js", "src/config.js", "src/helpers.js",
        "src/constants.js", "src/index.js", "tests/test.js",
        "docs/notes.md", "styles/main.css", "lib/core.js",
        "src/api.js"
    ];
    let filename = filenames[index % filenames.length];
    
    let filepath = path.join(REPO_DIR, filename);
    fs.mkdirSync(path.dirname(filepath), { recursive: true });
    fs.writeFileSync(filepath, content);
    
    let env = Object.assign({}, process.env, {
        GIT_AUTHOR_DATE: dateStr,
        GIT_COMMITTER_DATE: dateStr
    });
    
    try {
        execSync('git add .', { cwd: REPO_DIR, env });
        execSync(`git commit -m "${message}" --allow-empty-message`, { cwd: REPO_DIR, env, stdio: 'ignore' });
    } catch (e) {
        // ignore
    }
}

function main() {
    console.log("🎨 Generating realistic contribution history...");
    
    let totalCommits = 0;
    let totalActiveDays = 0;
    let current = new Date(START_DATE);
    
    // Setup git locally if needed, assuming already git init'd
    
    while (current <= END_DATE) {
        let count = getCommitCount(current);
        if (count > 0) {
            totalActiveDays++;
            for (let i = 0; i < count; i++) {
                createCommit(current, totalCommits + i);
            }
            totalCommits += count;
            
            if (totalActiveDays % 20 === 0) {
                console.log(`   📅 ${current.toISOString().split('T')[0]} | ${totalCommits} commits so far...`);
            }
        }
        current.setDate(current.getDate() + 1);
    }
    
    console.log(`\n✅ Done! Generated ${totalCommits} commits across ${totalActiveDays} active days`);
}

main();

File: generate_contributions.py
#!/usr/bin/env python3
"""
Generates realistic-looking GitHub contribution history via backdated git commits.
Creates a pattern that looks like a real final-year CS student:
- More active on weekdays, lighter on weekends
- Occasional bursts (hackathons, project deadlines)  
- Some completely empty weeks (exams, vacations)
- Varying intensity throughout the year
"""

import os
import random
import subprocess
from datetime import datetime, timedelta

# ─── Configuration ───────────────────────────────────────────────────
REPO_DIR = os.path.dirname(os.path.abspath(__file__))
START_DATE = datetime(2025, 7, 1)   # ~1 year of history
END_DATE = datetime(2026, 7, 20)    # Up to yesterday

# Seasonal activity weights (month -> base probability of committing)
# Simulates exam periods, vacations, project crunch times
MONTH_WEIGHTS = {
    1: 0.30,   # Jan: Exam season, low activity
    2: 0.50,   # Feb: Post-exams, moderate
    3: 0.65,   # Mar: Project season picking up
    4: 0.70,   # Apr: Active development
    5: 0.35,   # May: End-sem exams
    6: 0.25,   # Jun: Summer break start (low)
    7: 0.55,   # Jul: Summer projects / internship
    8: 0.60,   # Aug: Back to work
    9: 0.70,   # Sep: Semester projects active
    10: 0.75,  # Oct: Hackathon season, very active
    11: 0.65,  # Nov: Project submissions
    12: 0.30,  # Dec: End-sem exams + holiday
}

# Day-of-week weights (0=Monday, 6=Sunday)
DAY_WEIGHTS = {
    0: 0.80,   # Monday
    1: 0.85,   # Tuesday (most productive)
    2: 0.80,   # Wednesday
    3: 0.75,   # Thursday
    4: 0.65,   # Friday (winding down)
    5: 0.35,   # Saturday (occasional)
    6: 0.20,   # Sunday (rare)
}

# Define "burst weeks" — simulate hackathons/deadlines with higher activity
BURST_PERIODS = [
    (datetime(2025, 10, 10), datetime(2025, 10, 18)),   # Hackathon
    (datetime(2025, 11, 20), datetime(2025, 11, 30)),   # Project deadline
    (datetime(2026, 2, 5), datetime(2026, 2, 15)),      # New project kickoff
    (datetime(2026, 3, 15), datetime(2026, 3, 25)),     # Major feature push
    (datetime(2026, 4, 10), datetime(2026, 4, 20)),     # Alerion AI sprint
    (datetime(2026, 7, 1), datetime(2026, 7, 15)),      # Recent active period
]

# Define "dead zones" — exam weeks, vacations with zero commits
DEAD_PERIODS = [
    (datetime(2025, 12, 1), datetime(2025, 12, 20)),    # End-sem exams
    (datetime(2025, 12, 24), datetime(2026, 1, 3)),     # Holiday break
    (datetime(2026, 1, 5), datetime(2026, 1, 20)),      # Exam prep
    (datetime(2026, 5, 1), datetime(2026, 5, 20)),      # End-sem exams
    (datetime(2026, 6, 5), datetime(2026, 6, 18)),      # Break
]

# Commit message templates for realism
COMMIT_MESSAGES = [
    "refactor: clean up utility functions",
    "feat: add input validation",
    "fix: resolve edge case in parser",
    "docs: update API documentation",
    "style: improve code formatting",
    "test: add unit tests for auth module",
    "chore: update dependencies",
    "feat: implement caching layer",
    "fix: handle null pointer exception",
    "refactor: extract helper methods",
    "feat: add error handling middleware",
    "docs: add inline comments",
    "fix: correct off-by-one error",
    "feat: implement pagination",
    "chore: clean up unused imports",
    "test: add integration tests",
    "feat: add logging infrastructure",
    "fix: resolve race condition",
    "refactor: optimize database queries",
    "feat: implement rate limiting",
    "docs: update README with examples",
    "fix: handle timeout errors gracefully",
    "feat: add configuration validation",
    "style: standardize naming conventions",
    "test: improve test coverage",
    "feat: add health check endpoint",
    "fix: memory leak in event handler",
    "refactor: simplify control flow",
    "chore: configure CI pipeline",
    "feat: implement WebSocket handler",
    "fix: correct data serialization",
    "docs: add architecture diagram",
    "feat: add search functionality",
    "test: add edge case tests",
    "refactor: modularize components",
    "feat: implement data export",
    "fix: resolve CORS issues",
    "chore: update build configuration",
    "feat: add user preferences",
    "fix: handle concurrent requests",
]

# Dummy file content templates
FILE_CONTENTS = [
    "// Updated: {date}\n// Module improvements\nconst VERSION = '{version}';\n",
    "# Config updated {date}\nDEBUG={debug}\nLOG_LEVEL=info\n",
    "/* Style update {date} */\n.container {{ margin: {margin}px; }}\n",
    "# Notes - {date}\n- Refactored module\n- Fixed {bugs} bugs\n",
    "// Test suite - {date}\ndescribe('module', () => {{ /* tests */ }});\n",
]


def is_in_period(date, periods):
    """Check if date falls within any of the given periods."""
    for start, end in periods:
        if start <= date <= end:
            return True
    return False


def get_commit_count(date):
    """Determine how many commits to make on a given date."""
    # Dead periods = 0 commits
    if is_in_period(date, DEAD_PERIODS):
        return 0

    month_weight = MONTH_WEIGHTS.get(date.month, 0.5)
    day_weight = DAY_WEIGHTS.get(date.weekday(), 0.5)

    # Burst periods boost activity
    if is_in_period(date, BURST_PERIODS):
        month_weight = min(month_weight * 1.8, 1.0)
        day_weight = min(day_weight * 1.5, 1.0)

    # Combined probability
    probability = month_weight * day_weight

    # Add some randomness
    probability *= random.uniform(0.6, 1.4)

    # Roll the dice
    if random.random() > probability:
        return 0

    # Determine commit count based on intensity
    if is_in_period(date, BURST_PERIODS):
        # Burst: 1-7 commits
        weights = [15, 25, 25, 15, 10, 5, 5]
        return random.choices(range(1, 8), weights=weights)[0]
    else:
        # Normal: 1-4 commits, heavily weighted toward 1-2
        weights = [45, 30, 15, 10]
        return random.choices(range(1, 5), weights=weights)[0]


def create_commit(date, index):
    """Create a single backdated commit."""
    # Random time during the day (weighted toward evening for a student)
    hour_weights = {
        range(0, 6): 5,     # Late night
        range(6, 9): 5,     # Early morning (rare)
        range(9, 12): 15,   # Morning
        range(12, 14): 10,  # Lunch
        range(14, 18): 25,  # Afternoon
        range(18, 22): 30,  # Evening (peak)
        range(22, 24): 10,  # Night
    }

    # Pick hour with weights
    hours = []
    weights = []
    for hr_range, weight in hour_weights.items():
        for h in hr_range:
            hours.append(h)
            weights.append(weight)

    hour = random.choices(hours, weights=weights)[0]
    minute = random.randint(0, 59)
    second = random.randint(0, 59)

    commit_date = date.replace(hour=hour, minute=minute, second=second)
    date_str = commit_date.strftime("%Y-%m-%dT%H:%M:%S")

    # Pick a random commit message
    message = random.choice(COMMIT_MESSAGES)

    # Modify a file to create a real diff
    content_template = random.choice(FILE_CONTENTS)
    content = content_template.format(
        date=date.strftime("%Y-%m-%d"),
        version=f"{random.randint(1,9)}.{random.randint(0,9)}.{random.randint(0,99)}",
        debug=random.choice(["true", "false"]),
        margin=random.randint(4, 32),
        bugs=random.randint(1, 12),
    )

    # Use rotating file names so diffs are varied
    filenames = [
        "src/utils.js", "src/config.js", "src/helpers.js",
        "src/constants.js", "src/index.js", "tests/test.js",
        "docs/notes.md", "styles/main.css", "lib/core.js",
        "src/api.js",
    ]
    filename = filenames[index % len(filenames)]

    filepath = os.path.join(REPO_DIR, filename)
    os.makedirs(os.path.dirname(filepath), exist_ok=True)

    with open(filepath, "w") as f:
        f.write(content)

    env = os.environ.copy()
    env["GIT_AUTHOR_DATE"] = date_str
    env["GIT_COMMITTER_DATE"] = date_str

    subprocess.run(["git", "add", "."], cwd=REPO_DIR, capture_output=True)
    subprocess.run(
        ["git", "commit", "-m", message, "--allow-empty-message"],
        cwd=REPO_DIR,
        env=env,
        capture_output=True,
    )


def main():
    print("🎨 Generating realistic contribution history...")
    print(f"   Period: {START_DATE.strftime('%Y-%m-%d')} → {END_DATE.strftime('%Y-%m-%d')}")
    print()

    total_commits = 0
    total_active_days = 0
    current = START_DATE

    while current <= END_DATE:
        count = get_commit_count(current)

        if count > 0:
            total_active_days += 1
            for i in range(count):
                create_commit(current, total_commits + i)
            total_commits += count

            # Progress indicator
            if total_active_days % 20 == 0:
                print(f"   📅 {current.strftime('%Y-%m-%d')} | {total_commits} commits so far...")

        current += timedelta(days=1)

    total_days = (END_DATE - START_DATE).days
    print()
    print(f"✅ Done! Generated {total_commits} commits across {total_active_days} active days")
    print(f"   📊 {total_days} total days | {total_active_days} active | {total_days - total_active_days} inactive")
    print(f"   📈 Average: {total_commits / max(total_active_days, 1):.1f} commits per active day")
    print()
    print("🚀 Now push with: git push -u origin main")


if __name__ == "__main__":
    random.seed(2409)  # Fixed seed for reproducibility
    main()
